pkg/proxy/iptables/proxier_test.go

```diff
@@ -1345,6 +1347,270 @@ func assertIPTablesRulesNotEqual(t *testing.T, expected, result string) {
+	}
+
+	return tracer, nil
+}
+
+// ruleMatches checks if the give iptables rule matches (at least probabilistically) a
+// packet with the given sourceIP, destIP, and destPort.
+func (tracer *iptablesTracer) ruleMatches(rule, sourceIP, destIP, destPort string) bool {
+	var match []string
+
+	rule = commentRegex.ReplaceAllString(rule, "")
+
+	for _, moduleMatches := range moduleRegex.FindAllStringSubmatch(rule, -1) {
+		if noMatchModules.Has(moduleMatches[1]) {
+			return false
+		} else if !knownModules.Has(moduleMatches[1]) {
+			klog.ErrorS(nil, "Rule uses unknown iptables module; assuming no match", "rule", rule, "module", moduleMatches[1])
+			return false
+		}
+	}
+
+	// The sub-rules within an iptables rule are ANDed together, so the rule only
+	// matches if all of them match. So go through the subrules, and if any of them
+	// don't match, then fail.
+
+	// Match local/non-local
+	match = localRegex.FindStringSubmatch(rule)
+	if match != nil {
+		sourceIsLocal := sourceIP == tracer.nodeIP || sourceIP == "127.0.0.1"
+		destIsLocal := destIP == tracer.nodeIP || destIP == "127.0.0.1"
+
+		if match[1] == "" && match[2] == "src" && !sourceIsLocal {
+			return false
+		}
+		if match[1] == "!" && match[2] == "src" && sourceIsLocal {
+			return false
+		}
+		if match[1] == "" && match[2] == "dst" && !destIsLocal {
+			return false
+		}
+		if match[1] == "!" && match[2] == "dst" && destIsLocal {
+			return false
+		}
+	}
+
+	// Match destination IP/port
+	match = destIPRegex.FindStringSubmatch(rule)
+	if match != nil {
+		if match[1] == "!" && cidrContains(match[2], destIP) {
+			return false
+		} else if match[1] == "" && !cidrContains(match[2], destIP) {
+			return false
+		}
+	}
+	match = destPortRegex.FindStringSubmatch(rule)
+	if match != nil && match[1] != destPort {
+		return false
+	}
+
+	// Match source IP
+	match = sourceIPRegex.FindStringSubmatch(rule)
+	if match != nil {
+		if match[1] == "!" && cidrContains(match[2], sourceIP) {
+			return false
+		} else if match[1] == "" && !cidrContains(match[2], sourceIP) {
+			return false
+		}
+	}
+
+	// Ignore SessionAffinity rules
+	if affinityRegex.MatchString(rule) {
+		return false
+	}
+
+	// Anything else is assumed to match
+	return true
+}
+
+// runChain runs the given packet through the rules in the given table and chain. It
+// returns true if it hits a terminal action
+func (tracer *iptablesTracer) runChain(table, chain, sourceIP, destIP, destPort string) bool {
+	for _, rule := range tracer.tables[table][chain] {
+		match := jumpRegex.FindStringSubmatch(rule)
+		if match == nil {
+			continue
+		}
+
+		if !tracer.ruleMatches(rule, sourceIP, destIP, destPort) {
+			continue
+		}
+		tracer.matches = append(tracer.matches, rule)
+
+		switch match[1] {
+		case "KUBE-MARK-MASQ":
+			tracer.markMasq = true
+			continue
+
+		case "KUBE-MARK-DROP":
+			tracer.markDrop = true
+			continue
+
+		case "ACCEPT", "REJECT":
+			tracer.outputs = append(tracer.outputs, match[1])
```

Comment:
It's hard to follow the structure of these matches, could be helpful to have a comment indicated what `match[1]` represents here.